server/data/data-jmap/src/main/java/org/apache/james/jmap/api/upload/UploadServiceDefaultImpl.java

```diff
@@ -20,34 +20,106 @@
+        this.secureRandom = new SecureRandom();
     }
 
     @Override
     public Publisher<UploadMetaData> upload(InputStream data, ContentType contentType, Username user) {
-        return repository.upload(data, contentType, user);
+        return Mono.from(uploadRepository.upload(data, contentType, user))
+            .flatMap(upload -> Mono.from(uploadUsageRepository.increaseSpace(user, QuotaSizeUsage.size(upload.sizeAsLong())))
+                .thenReturn(upload))
+            .doOnSuccess(uploaded -> cleanupUploadIfNeeded(user)
+                .subscribeOn(ReactorUtils.BLOCKING_CALL_WRAPPER)
+                .subscribe());
     }
 
     @Override
     public Publisher<Upload> retrieve(UploadId id, Username user) {
-        return repository.retrieve(id, user);
+        return uploadRepository.retrieve(id, user);
+    }
+
+    private Mono<Void> cleanupUploadIfNeeded(Username username) {
+        return Mono.from(uploadUsageRepository.getSpaceUsage(username))
+            .map(QuotaSizeUsage::asLong)
+            .filter(quotaExceededPredicate())
+            .switchIfEmpty(Mono.empty())
+            .flatMap(quotaExceeded -> cleanupUpload(username, quotaExceeded));
+    }
+
+    private Predicate<Long> quotaExceededPredicate() {
+        return currentUploadUsage -> currentUploadUsage > jmapUploadQuotaConfiguration.getUploadQuotaLimitInBytes();
+    }
+
+    private Mono<Void> cleanupUpload(Username username, long currentUploadUsage) {
+        long minimumSpaceToClean = currentUploadUsage - jmapUploadQuotaConfiguration.getUploadQuotaLimitInBytes() / 2;
+        AtomicLong cleanedSpace = new AtomicLong(0L);
+
+        if (shouldPerformReadRepair()) {
+            return Flux.from(uploadRepository.listUploads(username))
+                .sort(Comparator.comparing(UploadMetaData::uploadDate))
+                .flatMap(upload -> {
+                    if (cleanedSpace.get() < minimumSpaceToClean) {
+                        return Mono.from(uploadRepository.delete(upload.uploadId(), username))
+                            .then(Mono.from(uploadUsageRepository.decreaseSpace(username, QuotaSizeUsage.size(upload.sizeAsLong()))))
+                            .then(Mono.fromCallable(() -> cleanedSpace.addAndGet(upload.sizeAsLong())))
+                            .then(Mono.empty());
+                    } else {
+                        return Mono.fromCallable(() -> upload);
+                    }
+                }, LOW_CONCURRENCY)
+                .map(UploadMetaData::sizeAsLong)
+                .reduce(Long::sum)
+                .filter(totalSpaceUsed -> totalSpaceUsed != currentUploadUsage)
+                .flatMap(totalSpaceUsed -> Mono.from(uploadUsageRepository.resetSpace(username,
+                    QuotaSizeUsage.size(totalSpaceUsed))))
+                .then();
+        } else {
+            return Flux.from(uploadRepository.listUploads(username))
+                .sort(Comparator.comparing(UploadMetaData::uploadDate))
```

Comment:
>  think the code WITHOUT read repair is somehow still better. Because it will terminate iterating the uploads list as soon as the 50% space condition is met. Meanwhile, the code WITH read repair still needs to iterate through all the uploads list to calculate the usage sum.

That's a non difference for the DB layer (read 5000 item batch ahead anyway)